src/librustc_typeck/check/method/probe.rs

```diff
@@ -837,9 +839,10 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
             for trait_candidate in applicable_traits.iter() {
                 let trait_did = trait_candidate.def_id;
                 if duplicates.insert(trait_did) {
-                    let import_id = trait_candidate.import_id.map(|node_id|
-                        self.fcx.tcx.hir().node_to_hir_id(node_id));
-                    let result = self.assemble_extension_candidates_for_trait(import_id, trait_did);
+                    let import_ids = trait_candidate.import_ids.iter().map(|node_id|
+                        self.fcx.tcx.hir().node_to_hir_id(*node_id)).collect();
```

Comment:
I think you can avoid this re-collection and just store `SmallVec<[NodeId; 1]>` in `Pick` and `Candidate`.
(And use `local_def_id` instead of `local_def_id_from_hir_id` when marking the IDs as used.)

Reply:
Disregard this.
The re-collection is suboptimal, but it was introduced as a part of `HirId`-ification and should go away when `TraitCandidate` is `HirId`-ified.